src/dmd/dsymbolsem.d

```diff
@@ -2803,6 +2803,30 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
         if (!sc)
             return;
 
+        if (ns.identExp)
+        {
+            // resolve name
+            sc = sc.startCTFE();
+            Expression resolved = ns.identExp.expressionSemantic(sc);
+            resolved = resolved.expressionSemantic(sc);
+            resolved = resolveProperties(sc, resolved);
+            sc = sc.endCTFE();
+            resolved = resolved.ctfeInterpret();
+            StringExp name = resolved.toStringExp();
+            if (!name)
+            {
+                ns.error("expected string expression for namespace name, got `%s`", ns.identExp.toChars());
+                return;
+            }
+            const(char)[] ident = name.toStringz();
+            if (!Identifier.isValidIdentifier(ident))
+            {
+                ns.error("expected valid identifer for C++ namespace but got `%.*s`", cast(int)ident.length, ident.ptr);
+                return;
+            }
+            ns.ident = Identifier.idPool(ident);
+        }
```

Comment:
Can I get guidance on this block? I'm not sure where/how it should be placed inside this function... and is there anything that should be done after the late-assignment of the symbol name that might have been missed prior?

Reply:
Please spilt that loop into the else case followed by a loop over the rest of the namespaces.

Reply:
So, the AST at the start looks like 
```
NSpace foo::bar::baz
+- Dsymbols members
```
After the end of the above it should look like (modulo `s.addMember(sc, sc.scopesym)`)
```
NSpace foo
+-Dsymbols members
+- NSpace bar
 +- NSpace baz
```
and you want it to end up looking like 
```
NSpace foo
+- NSpace bar
 +- NSpace baz
  +- Dsymbols members
```
In which case instead of that I'd validate all the namespace identifiers _first_, and then pull the ol' switcharoo:
```
Enclosing Scope (e.g. module)
+-NSpace foo::bar::baz
 +- Dsymbols members
```
split,  do _not_ insert into enclosing scope ->
```
NSpace bar::baz (parentless)
Enclosing Scope
+- NSpace foo
 +- Dsymbols members
```
expand ->
```
NSpace bar (parentless)
+- NSpace baz
Enclosing Scope
+- NSpace foo
 +- Dsymbols members
```
rotate reassign `ident`s ->
```
NSpace foo (parentless)
+- NSpace bar
Enclosing Scope
+- NSpace baz 
 +- Dsymbols members
```
pull the switcharoo (reparenting insert) ->
```
Enclosing Scope
+- NSpace foo
 +- NSpace bar
  +-NSpace baz 
   +- Dsymbols members
```

This means that the original namespace keeps all the members and the only reference to the original (because its a tree), from the enclosing scope, is overwritten. then you can `addMember` the intervening namespaces.

Reply:
Right... see, logically it makes sense... but i have no idea what the code that implements that might look like.
Those functions `addMember`, `setScope`, `importAll`, and what they mean relative to the 'current' scope (which is a concept that still isn't clear), and when they should all be called... and why. I can't imagine how to implement that sequence of operations with the primitives at hand.

Reply:
Not sure, try asking on slack.

Reply:
Please feel free to submit a patch if you think you know how to implement this logic ;)